lib/docker/util.rb

```diff
@@ -259,21 +284,22 @@ def build_config_header(credentials)
     }
   end
 
-  def glob_all_files(pattern)
-    Dir.glob(pattern, File::FNM_DOTMATCH) - ['..', '.']
+  # do a directory glob that matches .dockerignore behavior
+  # specifically: matched directories are considered a recursive match
+  def dockerignore_compatible_glob(pattern)
+    begin
+      some_dirs, some_files = glob_all_files(pattern).partition { |f| File.directory?(f) }
+      # since all directories will be re-processed with a /**/* glob, we can preemptively
+      # eliminate any whose parent directory is already in this set.  This saves significant time.
+      some_files + some_dirs.reject { |d| some_dirs.any? { |pd| d.start_with?(pd) && d != pd } }
+    end.each_with_object(Set.new) do |f, acc|
+      # expand any directories by globbing; flatten results
+      acc.merge(File.directory?(f) ? glob_all_files("#{f}/**/*") : [f])
+    end
   end
 
-  def remove_ignored_files!(directory, files)
-    ignore = File.join(directory, '.dockerignore')
-    return unless files.include?(ignore)
-    ignored_files(directory, ignore).each { |f| files.delete(f) }
+  def glob_all_files(pattern)
+    (Dir.glob(pattern, File::FNM_DOTMATCH) - ['..', '.']).reject { |p| p.end_with?("/.") }
```

Comment:
Is the change for this function needed? Specifically the reject for lines ending in `/.`. Just wondering when this would happen.

Reply:
Sure -- just added some debugging lines to find out where this happens.  

For a glob of `/Users/ianfixes/Development/docker-api/spec/docker/../fixtures/top/**/*`, one of the returned values is `/Users/ianfixes/Development/docker-api/spec/docker/../fixtures/top/.`